src/lib/stack.js

```diff
@@ -5,52 +5,78 @@ import 'babel-polyfill'
 
 import {
   UnavailableStackException,
-  UnavailableSettingsException
+  UnavailableSettingsException,
+  UnauthorizedStackException
 } from './exceptions'
 
 // the option credentials:include tells fetch to include the cookies in the
 // request even for cross-origin requests
-const fetchOptions = {
-  credentials: 'include'
+function fetchOptions () {
+  return {
+    credentials: 'include',
+    headers: {
+      Authorization: `Bearer ${COZY_TOKEN}`
+    }
+  }
 }
 
 let COZY_URL = __SERVER__
+let COZY_TOKEN
 
 async function getApps () {
-  try {
-    const response = await fetch(`${COZY_URL}/apps/`, fetchOptions)
-    const data = await response.json()
-    return data.data
-  } catch (e) {
+  const res = await fetch(`${COZY_URL}/apps/`, fetchOptions()).catch(e => {
     throw new UnavailableStackException()
+  })
+
+  if (res.status === 401) {
+    throw new UnauthorizedStackException()
   }
+
+  return (await res.json()).data
 }
 
 async function getDiskUsage () {
-  try {
-    const response = await fetch(`${COZY_URL}/settings/disk-usage`, fetchOptions)
-    const data = await response.json()
-    return data.data
-  } catch (e) {
+  const res = await fetch(`${COZY_URL}/settings/disk-usage`, fetchOptions()).catch(e => {
     throw new UnavailableStackException()
+  })
+
+  if (res.status === 401) {
+    throw new UnauthorizedStackException()
   }
+
+  return parseInt((await res.json()).data.attributes.used, 10)
 }
 
 async function getApp (slug) {
   return (await getApps()).find(item => item.attributes.slug === slug)
 }
 
 async function hasApp (slug) {
-  return !!(await getApp(slug))
+  const app = await getApp(slug)
+  return !!(app && app.attributes.state === 'ready')
 }
 
 module.exports = {
-  init ({cozyURL}) {
+  init ({cozyURL, token}) {
     COZY_URL = `//${cozyURL}`
+    COZY_TOKEN = token
   },
   has: {
     async settings () {
-      return await hasApp('settings')
+      let hasSettings
+
+      try {
+        hasSettings = await hasApp('settings')
+      } catch (e) {
+        hasSettings = false
+        throw new UnavailableSettingsException()
+      }
+
+      if (!hasSettings) {
+        throw new UnavailableSettingsException()
```

Comment:
Good point, I agree about this specific case of the settings app, maybe you can add a comment to be sure to not generalise this way if a similar function is necessary for another app? What do you think?